### ISSUE TITLE
Null<T> should be considered T for completion and resolving symbols

### ISSUE THREAD

**Issue by dpeek** [TiVo/intellij-haxe#304]
For the source:

```
class Foo {
    var bar:Null<String>;
    public function new() {}
}

class Baz {
    static function main() {
        new Foo().bar.split('');
    }
}
```

Completions are not provided for `bar` and `split` is unresolved. For completion and resolving `Null<String>` should behave as `String`.

**Comment by eliasku** [comment]
+1

Issue for all generic typedefs:
```
var r:Ref<String>;
r.get().split('');
```

**Comment by EBatTiVo** [comment]
I'm confused as to why the language would allow a parameterized value to be used as a type, but it does:

http://try.haxe.org/#70874:
```haxe
class Test {
    static function main() {
        trace("Haxe is great!");
        var b:Null<String>;
        trace(b);
        trace(b.length);  // <--- fails at run-time; completes as a String.
        var split = b.split('');
        trace(b);
        var i = 1;
        trace(i);
    }
}
``` 


The compiler seems to determine the type as a String, and try.haxe.org provides completion as if it were.  The analogous `var i:1;` doesn't compile.

**Comment by EBatTiVo** [comment]
Elias' example is different, because `r` is a `ref`.  So r.get() should return a String, and that means the parameter of the Ref<> isn't being resolved to a type.

**Comment by eliasku** [comment]
@EBatTiVo 

[Definition of Null type](http://haxe.org/manual/types-nullability.html) is `typedef Null<T> = T;`

*NOTICE: There are thoughts in Haxe community about changing `Null<T>` definition from typedef to abstract someday, or maybe not.*

`Null<String>` must be `String`,
`Null<Int>` must be `Int`.
`String` is naturally nullable by all targets, but `Int` is not, and `Null<Int>` is allow to set null to the variable, and it depends on target, for example C# has `struct MyStruct` types, which is always not-null, to make it nullable - there is `Null<MyStruct>`. Otherwise for JS target ALL std types is Nullable. But to unify API between targets - there is `Null<T>` construction.
`Null` type is very haxe-specific std feature, and make sense for compiler how to cross-compile code to specific target (python, c#, cpp, etc). But at least it has very clear definition `typedef Null<T> = T` for reading and interpretation. That is very easy to resolve, just pass Type to `T`, substitute it in typedef decl body and you'll get the way how to handle this variable.

I've made example for `Ref<T>` because of looks like any generic typedef declarations is not resolved at all. And `Ref<T>` is used much in type build macros.

More natural example how `Null<T>` is used:
```
function foo(users:Null<String>) { // actually equivalent of (?users:String)
    return users != null ? users.split(',') : '';
}
```
So another usage of `Null<T>` is just to annotate Nullability for method argument :)

I think I don't get right the `i:1` expression from you example. I'll try to annotate line-by-line. Let me know if I missed something:
```
class Test {
    static function main() {
        trace("Haxe is great!");
        var b:Null<String>; // b:String = null
        trace(b); // trace(null);
        trace(b.length);  // Null-pointer exception - b is null
        var split = b.split(''); // Null-pointer exception - b is null
        trace(b); // trace(null);
        var i = 1; // var i:Int = 1 - http://haxe.org/manual/expression-var.html
        trace(i); // trace(i); // 1
    }
}
```
In some 3.3.0 dev versions of Haxe you even will get compile error because `b` is not initialized before usage. (like it's done in java or c# compilers)

**Comment by EBatTiVo** [comment]
Fixed: https://github.com/TiVo/intellij-haxe/commit/8cdf7a9023eccf5f848325d1b3aa016556b483cd

Will be included in release 0.9.10.  Please reopen if it recurs.
